File: python-algo/gamelib/game_map.py
import math
from .unit import GameUnit
from .util import debug_write

class GameMap:
    """Holds data about the current game map and provides functions
    useful for getting information related to the map.

    game_map[x, y] will return a list of Units located at that location, 
    or an empty list if there are no units at the location

    Attributes:
        * config (JSON): Contains information about the game
        * ARENA_SIZE (int): The size of the arena.
        * HALF_ARENA (int): Half of the size of the arena.
        * TOP_RIGHT (int): A constant that represents the top right edge
        * TOP_LEFT (int): A constant that represents the top left edge
        * BOTTOM_LEFT (int): Hidden challenge! Can you guess what this constant represents???
        * BOTTOM_RIGHT (int): A constant that represents the bottom right edge

    """
    def __init__(self, config):
        """Initializes constants and game map

        Args:
            * config (JSON): Contains information about the game

        """
        self.config = config
        self.enable_warnings = True
        self.ARENA_SIZE = 28
        self.HALF_ARENA = int(self.ARENA_SIZE / 2)
        self.TOP_RIGHT = 0
        self.TOP_LEFT = 1
        self.BOTTOM_LEFT = 2
        self.BOTTOM_RIGHT = 3
        self.__map = self.__empty_grid()
        self.__start = [13,0]
    
    def __getitem__(self, location):
        if len(location) == 2 and self.in_arena_bounds(location):
            x,y = location
            return self.__map[x][y]
        self._invalid_coordinates(location)

    def __setitem__(self, location, val):
        if type(location) == tuple and len(location) == 2 and self.in_arena_bounds(location):
            self.__map[location[0]][location[1]] = val
            return
        self._invalid_coordinates(location)

    def __iter__(self):
        self.__start = [13,0]
        return self
    
    def __next__(self):
        location = self.__start
        if location == [15,27]:
            raise StopIteration
        new_location = [location[0]+1, location[1]]
        while not self.in_arena_bounds(new_location) and not location == [14,27]:
            if new_location[0] == self.ARENA_SIZE:
                new_location = [0, new_location[1]+1]
            else:
                new_location = [new_location[0]+1, new_location[1]]
        self.__start = new_location
        return location 

    def __empty_grid(self):
        grid = []
        for x in range(0, self.ARENA_SIZE):
            grid.append([])
            for _ in range(0, self.ARENA_SIZE):
                grid[x].append([])
        return grid

    def _invalid_coordinates(self, location):
        self.warn("{} is out of bounds.".format(str(location)))

    def in_arena_bounds(self, location):
        """Checks if the given location is inside the diamond shaped game board.

        Args:
            * location: A map location

        Returns:
            True if the location is on the board, False otherwise
        
        """
        x, y = location
        half_board = self.HALF_ARENA

        row_size = y + 1
        startx = half_board - row_size
        endx = startx + (2 * row_size) - 1
        top_half_check = (y < self.HALF_ARENA and x >= startx and x <= endx)

        row_size = (self.ARENA_SIZE - 1 - y) + 1
        startx = half_board - row_size
        endx = startx + (2 * row_size) - 1
        bottom_half_check = (y >= self.HALF_ARENA and x >= startx and x <= endx)

        return bottom_half_check or top_half_check

    def get_edge_locations(self, quadrant_description):
        """Takes in an edge description and returns a list of locations.
        
        Args:
            * quadrant_description: A constant corresponding to an edge. Valid quadrant descriptions are
                * GameMap.TOP_RIGHT
                * GameMap.TOP_LEFT
                * GameMap.BOTTOM_RIGHT
                * GameMap.BOTTOM_LEFT

        Returns:
            A list of locations corresponding to the requested edge

        """
        if not quadrant_description in [self.TOP_LEFT, self.TOP_RIGHT, self.BOTTOM_LEFT, self.BOTTOM_RIGHT]:
            self.warn("Passed invalid quadrant_description '{}'. See the documentation for valid inputs for get_edge_locations.".format(quadrant_description))
            return

        edges = self.get_edges()
        return edges[quadrant_description]

    def get_edges(self):
        """Gets all of the edges and their edge locations

        Returns:
            A list with four lists inside of it of locations corresponding to the four edges.
            [0] = top_right, [1] = top_left, [2] = bottom_left, [3] = bottom_right.
        """
        top_right = []
        for num in range(0, self.HALF_ARENA):
            x = self.HALF_ARENA + num
            y = self.ARENA_SIZE - 1 - num
            top_right.append([int(x), int(y)])
        top_left = []
        for num in range(0, self.HALF_ARENA):
            x = self.HALF_ARENA - 1 - num
            y = self.ARENA_SIZE - 1 - num
            top_left.append([int(x), int(y)])
        bottom_left = []
        for num in range(0, self.HALF_ARENA):
            x = self.HALF_ARENA - 1 - num
            y = num
            bottom_left.append([int(x), int(y)])
        bottom_right = []
        for num in range(0, self.HALF_ARENA):
            x = self.HALF_ARENA + num
            y = num
            bottom_right.append([int(x), int(y)])
        return [top_right, top_left, bottom_left, bottom_right]
    
    def add_unit(self, unit_type, location, player_index=0):
        """Add a single GameUnit to the map at the given location.

        Args:
            * unit_type: The type of the new unit
            * location: The location of the new unit
            * player_index: The index corresponding to the player controlling the new unit, 0 for you 1 for the enemy

        This function does not affect your turn and only changes the data stored in GameMap. The intended use of this function
        is to allow you to create arbitrary gamestates. Using this function on the GameMap inside game_state can cause your algo to crash.
        """
        if not self.in_arena_bounds(location):
            self._invalid_coordinates(location)
        if player_index < 0 or player_index > 1:
            self.warn("Player index {} is invalid. Player index should be 0 or 1.".format(player_index))

        x, y = location
        new_unit = GameUnit(unit_type, self.config, player_index, None, location[0], location[1])
        if not new_unit.stationary:
            self.__map[x][y].append(new_unit)
        else:
            self.__map[x][y] = [new_unit]

    def remove_unit(self, location):
        """Remove all units on the map in the given location.

        Args:
            * location: The location that you will empty of units

        This function does not affect your turn and only changes the data stored in GameMap. The intended use of this function
        is to allow you to create arbitrary gamestates. Using this function on the GameMap inside game_state can cause your algo to crash.
        """
        if not self.in_arena_bounds(location):
            self._invalid_coordinates(location)
        
        x, y = location
        self.__map[x][y] = []

    def get_locations_in_range(self, location, radius):
        """Gets locations in a circular area around a location

        Args:
            * location: The center of our search area
            * radius: The radius of our search area

        Returns:
            The locations that are within our search area

        """
        if radius < 0 or radius > self.ARENA_SIZE:
            self.warn("Radius {} was passed to get_locations_in_range. Expected integer between 0 and {}".format(radius, self.ARENA_SIZE))
        if not self.in_arena_bounds(location):
            self._invalid_coordinates(location)

        x, y = location
        locations = []
        for i in range(int(x - radius), int(x + radius + 1)):
            for j in range(int(y - radius), int(y + radius + 1)):
                new_location = [i, j]
                # A unit with a given range affects all locations who's centers are within that range + 0.51 so we add 0.51 here
                if self.in_arena_bounds(new_location) and self.distance_between_locations(location, new_location) < radius + 0.51:
                    locations.append(new_location)
        return locations

    def distance_between_locations(self, location_1, location_2):
        """Euclidean distance

        Args:
            * location_1: An arbitrary location
            * location_2: An arbitrary location

        Returns:
            The euclidean distance between the two locations

        """
        x1, y1 = location_1
        x2, y2 = location_2

        return math.sqrt((x1 - x2)**2 + (y1 - y2)**2)

    def warn(self, message):
        if(self.enable_warnings):
            debug_write(message)

File: community/c++-algo/include/GameLib/src/gameMap.cpp
/*
Description: Implementations for the algoMap header.
Last Modified: 10 Apr 2019
Author: Isaac Draper, Ryan Draves
*/

#include "gameMap.h"

namespace terminal {

    using json11::Json;
    using std::vector;

    /// Constructor for GameMap which requires a configuration Json object.
    /// @param configuration A Json object containing information about the game.
    GameMap::GameMap(Json config) {
        this->config = config;
        this->verbosity = WARNING;
        createEmptyGrid();
    }

    /// Fills map with an emtpy grid of dimensions AREA_SIZE.
    void GameMap::createEmptyGrid() {
        map.resize(ARENA_SIZE);
        for (int x = 0; x < ARENA_SIZE; ++x)
            map.at(x).resize(ARENA_SIZE);
    }

    /// Checks if a position is inside the diamond shaped game board.
    /// @param x The x position.
    /// @param y The y position.
    bool GameMap::inArenaBounds(int x, int y) const {
        int rowSize = y + 1;
        int startX = HALF_ARENA - rowSize;
        int endX = startX + (2 * rowSize) - 1;
        const bool topHalfCheck = y < HALF_ARENA && x >= startX && x <= endX;

        rowSize = (ARENA_SIZE - 1 - y) + 1;
        startX = HALF_ARENA - rowSize;
        endX = startX + (2 * rowSize) - 1;
        const bool bottomHalfCheck = y >= HALF_ARENA && x >= startX && x <= endX;

        return bottomHalfCheck || topHalfCheck;
    }

    /// Checks if a position is inside the diamond shaped game board.
    /// @param pos The position to check.
    bool GameMap::inArenaBounds(Pos pos) const {
        return inArenaBounds(pos.x, pos.y);
    }

    /// Takes an edge and fills a vector with a list of locations.
    /// @param vec A vector passed by reference you would like to fill.
    /// @param edge The edge to get units for.
    void GameMap::getEdgeLocations(vector<Pos>& vec, const EDGE edge) const {
        int x, y;
        switch (edge) {
        case TOP_RIGHT: {
            for (int i = 0; i < HALF_ARENA; ++i) {
                x = HALF_ARENA + i;
                y = ARENA_SIZE - 1 - i;
                vec.push_back(Pos{ x, y });
            }
            break;
        }
        case TOP_LEFT: {
            for (int i = 0; i < HALF_ARENA; ++i) {
                x = HALF_ARENA - 1 - i;
                y = ARENA_SIZE - 1 - i;
                vec.push_back(Pos{ x, y });
            }
            break;
        }
        case BOTTOM_LEFT: {
            for (int i = 0; i < HALF_ARENA; ++i) {
                x = HALF_ARENA - 1 - i;
                y = i;
                vec.push_back(Pos{ x, y });
            }
            break;
        }
        case BOTTOM_RIGHT: {
            for (int i = 0; i < HALF_ARENA; ++i) {
                x = HALF_ARENA + i;
                y = i;
                vec.push_back(Pos{ x, y });
            }
            break;
        }
        default: {
                Util::printError<GameMapException>("Invalid edge requested", INVARIANT, verbosity);
        }
        }
    }

    /// Fills out a vector of all the edges
    /// @param vec A vector of { topRight, topLeft, bottomLeft, bottomRight } edges
    void GameMap::getEdges(vector<vector<Pos> >& vec) {
        vector<Pos> topRight, topLeft, bottomLeft, bottomRight;
        getEdgeLocations(topRight, TOP_RIGHT);
        getEdgeLocations(topLeft, TOP_LEFT);
        getEdgeLocations(bottomLeft, BOTTOM_LEFT);
        getEdgeLocations(bottomRight, BOTTOM_RIGHT);
        vec.emplace_back(topRight);
        vec.emplace_back(topLeft);
        vec.emplace_back(bottomLeft);
        vec.emplace_back(bottomRight);
    }

    /// Add a single GameUnit to the map at the given location.
    /// This does not send it to the engine, it simply lets you create any
    /// map position you want to experiment with.
    /// @param unitType The type of unit to add. Stationary units will replace 
    /// @param pos The position to add the unit at.
    /// @param playerIndex The player to add the unit for.
    /// @param hp The health of the unit (default is max health).
    void GameMap::addUnit(UNIT_TYPE unitType, Pos pos, int playerIndex, double hp) {
        if (!inArenaBounds(pos)) Util::printError<PosException>("Out of bounds exception", CRASH, verbosity);
        if (playerIndex < 0 || playerIndex > 1) throw PlayerIndexException();
        // Stability of 0 will default the unit to max_stability
        GameUnit newUnit = GameUnit(unitType, config, hp, playerIndex, pos[0], pos[1]);
        if (!newUnit.stationary) {
            map.at(pos.x).at(pos.y).push_back(newUnit);
        }
        else {
            size_t size = map.at(pos.x).at(pos.y).size();
            if (size > 0 && newUnit.unitType != REMOVE)
                Util::printError<GameMapException>("Error placing a stationary unit in an occupied location", INVARIANT, verbosity);
            else if ((size != 1 || (size > 0 && !map.at(pos.x).at(pos.y).at(0).stationary))
                && newUnit.unitType == REMOVE)
                Util::printError<GameMapException>("Error placing remove; 1 stationary unit not found", INVARIANT, verbosity);
            // Clearing the vector for firewalls is not necessary, as we verified above
            map.at(pos.x).at(pos.y).push_back(newUnit);
        }
    }


    /// Add a single GameUnit to the map at the given location.
    /// This does not send it to the engine, it simply lets you create any
    /// map position you want to experiment with.
    /// @param unitType The type of unit to add. Stationary units will replace 
    /// @param x The x position to add the unit at.
    /// @param y The y position to add the unit at.
    /// @param playerIndex The player to add the unit for.
    /// @param hp The health of the unit (default is max health).
    void GameMap::addUnit(UNIT_TYPE unitType, int x, int y, int playerIndex, double hp) {
        addUnit(unitType, Pos(x, y), playerIndex, hp);
    }

    /// Remove all GameUnits from the game map at the location. Throw an error if it's empty.
    /// @param pos Position to clear from the game map
    void GameMap::removeUnits(Pos pos) {
        if (!inArenaBounds(pos)) Util::printError<PosException>("Out of bounds exception", CRASH, verbosity);
        if (map.at(pos.x).at(pos.y).size() == 0) Util::printError<GameMapException>("Error trying to remove 0 units", WARNING, verbosity);
        map.at(pos.x).at(pos.y).clear();
    }

    /// Takes a position and radius and fills a vector with locations in its range
    /// @param locations A vector passed by reference to return the values.
    /// @param pos The position to find locatins in range from.
    /// @param radius The radius of the circle to get locations from.
    void GameMap::getLocationsInRange(vector<Pos>& locations, Pos pos, double radius) {
        if (radius < 0 || radius > ARENA_SIZE)
            Util::printError<GameMapException>("Error getting locations in range with that radius", INVARIANT, verbosity);
        if (!inArenaBounds(pos)) Util::printError<PosException>("Out of bounds exception", CRASH, verbosity);
        for (int i = (int)(pos.x - radius); i < (int)(pos.x + radius + 1); i++) {
            for (int j = (int)(pos.y - radius); j < (int)(pos.y + radius + 1); j++) {
                Pos new_pos = { i, j };
                if (inArenaBounds(new_pos) && distanceBetweenLocations(pos, new_pos) < radius + 0.51) {
                    locations.push_back(new_pos);
                }
            }
        }
    }

    /// Boolean return of a stationary unit located at the position.
    /// @param pos Position to check on the map.
    /// @return Boolean answer to a stationary unit at the location.
    bool GameMap::containsStationaryUnit(Pos pos) const {
        if (!inArenaBounds(pos)) Util::printError<PosException>("Out of bounds exception", CRASH, verbosity);
        if (map.at(pos.x).at(pos.y).size() > 0) {
            switch (map.at(pos.x).at(pos.y).at(0).unitType) {
                case FILTER:
                case ENCRYPTOR:
                case DESTRUCTOR:
                    return true;
                    break;
                default:
                    return false;
            }
        }
        return false;
    }

    /// Boolean return of a stationary unit located at the position.
    /// @param x X coordinate to check on the map.
    /// @param y Y coordinate to check on the map.
    /// @return Boolean answer to a stationary unit at the location.
    bool GameMap::containsStationaryUnit(int x, int y) const {
        return containsStationaryUnit(Pos(x, y));
    }

    /// Helper function to get the euclidean distance between two locations
    /// @param pos_1 First position.
    /// @param pos_2 Second position.
    /// @return Euclidean distance between two positions.
    double GameMap::distanceBetweenLocations(Pos pos_1, Pos pos_2) {
        return sqrt(pow(pos_1.x - pos_2.x, 2) + pow(pos_1.y - pos_2.y, 2));
    }

    /// Overloaded [Pos] operator enables access in the game map with a Pos.
    /// @param pos Position to index into the map.
    /// @return A reference to the vector of GameUnits at the location.
    vector<GameUnit>& GameMap::operator[](const Pos &pos) {
        if (!inArenaBounds(pos)) Util::printError<PosException>("Out of bounds exception", CRASH, verbosity);
        return map.at(pos.x).at(pos.y);
    }

    /// Overloaded const [Pos] operator enables const access in the game map with a Pos.
    /// @param pos Position to index into the map.
    /// @return A reference to the vector of GameUnits at the location.
    const vector<GameUnit>& GameMap::operator[](const Pos &pos) const {
        if (!inArenaBounds(pos)) Util::printError<PosException>("Out of bounds exception", CRASH, verbosity);
        return map.at(pos.x).at(pos.y);
    }

    /// Overloaded [int] operator enables access in the game map with [x][y]
    /// @param x X coordinate to index into the map.
    /// @return A reference to the column of the game map at coordinate X.
    vector<vector<GameUnit> >& GameMap::operator[](int x) {
        // y = 13 and 14 are true to any valid x value
        if (!inArenaBounds(Pos(x, 13))) Util::printError<PosException>("Out of bounds exception", CRASH, verbosity);
        return map.at(x);
    }

    /// Iterator function to help auto range through the game map
    /// @return A vector begin iterator through the columns of the game map
    vector<vector<vector<GameUnit> > >::iterator GameMap::begin() {
        return map.begin();
    }

    /// Iterator function to help auto range through the game map
    /// @return A vector end iterator through the columns of the game map
    vector<vector<vector<GameUnit> > >::iterator GameMap::end() {
        return map.end();
    }

    /// Prints an ACSII version of the current game map for debug purposes
    /// @return A very large string that represents the GameMap.
    string GameMap::toString() const {
        string ret = "";
        for (int y = 0; y < ARENA_SIZE; y++) {
            for (int x = 0; x < ARENA_SIZE; x++) {
                Pos location = { x, ARENA_SIZE - y - 1 };
                const vector<GameUnit> &stack = map.at(location.x).at(location.y);
                if (!inArenaBounds(location))
                    ret += " - ";
                else if (stack.size() == 0)
                    ret += " + ";
                else {
                    ret += stack.at(0).toString() + " ";
                }
            }
            ret += "\n";
        }
        return ret;
    }

}
